daemon/cluster/noderunner.go

```diff
@@ -0,0 +1,286 @@
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	n.reconnectDelay = initialReconnectDelay
+
+	return n.start(conf)
+}
+
+func (n *nodeRunner) start(conf nodeStartConfig) error {
+	var control string
+	if runtime.GOOS == "windows" {
+		control = `\\.\pipe\` + controlSocket
+	} else {
+		control = filepath.Join(n.cluster.runtimeRoot, controlSocket)
+	}
+
+	node, err := swarmnode.New(&swarmnode.Config{
+		Hostname:           n.cluster.config.Name,
+		ForceNewCluster:    conf.forceNewCluster,
+		ListenControlAPI:   control,
+		ListenRemoteAPI:    conf.ListenAddr,
+		AdvertiseRemoteAPI: conf.AdvertiseAddr,
+		JoinAddr:           conf.joinAddr,
+		StateDir:           n.cluster.root,
+		JoinToken:          conf.joinToken,
+		Executor:           container.NewExecutor(n.cluster.config.Backend),
+		HeartbeatTick:      1,
+		ElectionTick:       3,
+		UnlockKey:          conf.lockKey,
+		AutoLockManagers:   conf.autolock,
+	})
+	if err != nil {
+		return err
+	}
+	if err := node.Start(context.Background()); err != nil {
+		return err
+	}
+
+	n.done = make(chan struct{})
+	n.ready = make(chan struct{})
+	n.swarmNode = node
+	n.config = conf
+	savePersistentState(n.cluster.root, conf)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		err := detectLockedError(node.Err(context.Background()))
+		if err != nil {
+			logrus.Errorf("cluster exited with error: %v", err)
+		}
+		n.mu.Lock()
+		n.swarmNode = nil
+		n.err = err
+		close(n.done)
+		select {
+		case <-n.ready:
+			n.enableReconnectWatcher()
+		default:
+			if n.repeatedRun {
+				n.enableReconnectWatcher()
+			}
+		}
+		n.repeatedRun = true
+		n.mu.Unlock()
+		cancel()
+	}()
+
+	go func() {
+		select {
+		case <-node.Ready():
+			n.mu.Lock()
+			n.err = nil
+			n.mu.Unlock()
+			close(n.ready)
+		case <-ctx.Done():
+		}
+		n.cluster.configEvent <- struct{}{}
+	}()
+
+	go func() {
+		for conn := range node.ListenControlSocket(ctx) {
+			n.mu.Lock()
+			if n.grpcConn != conn {
+				if conn == nil {
+					n.controlClient = nil
+					n.logsClient = nil
+				} else {
+					n.controlClient = swarmapi.NewControlClient(conn)
+					n.logsClient = swarmapi.NewLogsClient(conn)
+				}
+			}
+			n.grpcConn = conn
+			n.mu.Unlock()
+			n.cluster.configEvent <- struct{}{}
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops the current swarm node if it is running.
+func (n *nodeRunner) Stop() error {
+	n.mu.Lock()
+	if n.cancelReconnect != nil { // between restarts
+		n.cancelReconnect()
+		n.cancelReconnect = nil
+	}
+	if n.swarmNode == nil {
+		n.mu.Unlock()
+		return nil
+	}
+	n.stopping = true
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+	if err := n.swarmNode.Stop(ctx); err != nil && !strings.Contains(err.Error(), "context canceled") {
+		n.mu.Unlock()
+		return err
+	}
+	n.mu.Unlock()
+	<-n.done
+	return nil
+}
+
+func (n *nodeRunner) State() NodeState {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+
+	ns := n.NodeState
```

Comment:
Don't we need to copy `n.NodeState` here?
I expect `State()` won't update any update field of `n`.